### ISSUE TITLE
关于GPLinker训练中训练会出现LOSS为负的问题

### ISSUE THREAD

**Issue by ZOLETH** [bojone/bert4keras#500]
python版本：3.6
tensorflow版本 2.6
keras版本 2.6
pytorch版本 1.10

我使用的是[xhw205]实现的pytorch版的GPlinker (https://github.com/xhw205/GPLinker_torch) ，当时他也提到训练过程中会出现LOSS为负值的情况，我排查后发现可能是sparse_multilabel_categorical_crossentropy的计算方式导致的，以下是[xhw205]实现的torch版的部分代码：
    all_loss = torch.logsumexp(y_pred, dim=-1)
    aux_loss = torch.logsumexp(y_pos_2, dim=-1) - all_loss
    aux_loss = torch.clip(1 - torch.exp(aux_loss), 1e-10, 1)
    neg_loss = all_loss + torch.log(aux_loss)
对应keras代码为：
    all_loss = K.logsumexp(y_pred, axis=-1)
    aux_loss = K.logsumexp(y_pos_2, axis=-1) - all_loss
    aux_loss = K.clip(1 - K.exp(aux_loss), K.epsilon(), 1)
    neg_loss = all_loss + K.log(aux_loss)
假设**y_pos_2 = [9.048806190490723, -inf], all_loss = [-inf, 9.048806190490723, -27.048492431640625, 0.0]**，按照https://kexue.fm/archives/8888 中公式的思路应该会得到一个值为正的neg_loss, 但实际上经过上述torch代码计算后会得到一个负的neg_loss，经过测试在**keras版也同样会计算出负值**，也就是这个负的neg_loss导致了最后的loss为负数。我实验后发现可能是框架函数实现上的问题，比如-27.048492431640625这个数经过np.exp和torch.exp计算得到的结果是不一样的(torch和keras的结果貌似是一样的)，这说明框架对类似exp和log这种操作可能没法得到一个非常精确的值，这种不精确在文章公式(4)→(5)的过程中经过大量连续log和exp操作被放大了，最后就形成了一个负值。

我觉得是否可以将neg_loss的计算改为如下形式：
neg_loss = K.log(K.sum(K.exp(y), axis=-1) - K.sum(K.exp(x), axis=-1))
即根据文章中的公式（4）直接在第一步就算出neg_loss, 至少这样改完后我在torch版的训练过程中就没有出现loss为负的情况了
